FastSimulation/SimplifiedGeometryPropagator/plugins/FastTrackDeDxProducer.cc

```diff
@@ -0,0 +1,207 @@
+
+  m_tracksTag = consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("tracks"));
+
+  //simHitsToken = consumes<edm::PSimHitContainer>(iConfig.getParameter<edm::InputTag>("simHits"));
+  //simHit2RecHitMapToken = consumes<FastTrackerRecHitRefCollection>(iConfig.getParameter<edm::InputTag>("simHit2RecHitMap"));
+
+  usePixel = iConfig.getParameter<bool>("UsePixel"); 
+  useStrip = iConfig.getParameter<bool>("UseStrip");
+  meVperADCPixel = iConfig.getParameter<double>("MeVperADCPixel"); 
+  meVperADCStrip = iConfig.getParameter<double>("MeVperADCStrip"); 
+
+  shapetest = iConfig.getParameter<bool>("ShapeTest");
+  useCalibration = iConfig.getParameter<bool>("UseCalibration");
+  m_calibrationPath = iConfig.getParameter<string>("calibrationPath");
+
+  convertFromGeV2MeV = iConfig.getParameter<bool>("convertFromGeV2MeV");
+  nothick = iConfig.getParameter<bool>("nothick");  
+
+  if(!usePixel && !useStrip)
+  	throw cms::Exception("fastsim::SimplifiedGeometry::FastTrackDeDxProducer.cc") << " Pixel Hits AND Strip Hits will not be used to estimate dEdx --> BUG, Please Update the config file";
+
+
+}
+
+
+FastTrackDeDxProducer::~FastTrackDeDxProducer()
+{
+  delete m_estimator;
+}
+
+// ------------ method called once each job just before starting event loop  ------------
+void  FastTrackDeDxProducer::beginRun(edm::Run const& run, const edm::EventSetup& iSetup)
+{
+  if(useCalibration && calibGains.empty()){
+    edm::ESHandle<TrackerGeometry> tkGeom;
+    iSetup.get<TrackerDigiGeometryRecord>().get( tkGeom );
+    m_off = tkGeom->offsetDU(GeomDetEnumerators::PixelBarrel); //index start at the first pixel
+	
+    DeDxTools::makeCalibrationMap(m_calibrationPath, *tkGeom, calibGains, m_off);
+  }
+
+  m_estimator->beginRun(run, iSetup);
+}
+
+
+
+void FastTrackDeDxProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+
+  auto trackDeDxEstimateAssociation = std::make_unique<ValueMap<DeDxData>>();  
+  ValueMap<DeDxData>::Filler filler(*trackDeDxEstimateAssociation);
+
+  edm::Handle<reco::TrackCollection> trackCollectionHandle;
+  iEvent.getByToken(m_tracksTag,trackCollectionHandle);
+  //std::vector<DeDxData> dedxEstimate( trackCollectionHandle->size() );
+  const auto& trackCollection = *trackCollectionHandle;
+  std::vector<DeDxData> dedxEstimate( trackCollection.size() );
+
+
+
+  for(unsigned int j=0;j<trackCollectionHandle->size();j++){            
+    const reco::TrackRef track = reco::TrackRef( trackCollectionHandle.product(), j );
+    
+    int NClusterSaturating = 0; 
+    DeDxHitCollection dedxHits;
+      
+    auto const & trajParams = track->extra()->trajParams();
+    assert(trajParams.size()==track->recHitsSize());
+      
+    auto hb = track->recHitsBegin();
+    dedxHits.reserve(track->recHitsSize()/2);
+    for(unsigned int h=0;h<track->recHitsSize();h++){
+	const FastTrackerRecHit recHit = static_cast< const FastTrackerRecHit & >(*(*(hb+h)));
```

Comment:
```suggestion
	const FastTrackerRecHit& recHit = static_cast< const FastTrackerRecHit & >(*(*(hb+h)));
```
to prevent unneeded copy

Reply:
Actually how about looping with the iterator?
```cpp
    for(; hb != track->recHitsEnd(); ++hb){
      const FastTrackerRecHit& recHit = static_cast< const FastTrackerRecHit & >(**hb);
```
?

Reply:
the line inside the for-loop:

auto trackDirection = trajParams[h].direction();

needs to have h, the simple integer iterator, in order to access the trajectory - not sure if this would work with the iterator choice suggested.

Reply:
Right, I missed that use of `h`, thanks for the clarification.